### ISSUE TITLE
[Bug?] Reruns of Scenario outlines lead to potentially confusing scenario names

### ISSUE THREAD

**Issue by timhunt** [Behat/Behat#1448]
To explain what is going on, I am going to use a silly example, just to keep it short. Suppose random.html is a pointless page which displays 1 or 2 at random, and we test it like this.

```cucumber
  Scenario Outline: Unreliable test
    When I load "random.html"
    Then I should see "<expected>"
    Examples:
      | expected |
      | 1        |
      | 2        |
```

Suppose that we run this, and get the results:
`Unreliable test #1` -> PASS
`Unreliable test #2` -> FAIL

Suppose we are using the --rerun option to the test-runner. When we do the re-run, it correctly only runs the second example. However, in the result, it names it #1. That is

`Unreliable test #1` -> PASS (this did correctly check for and find 2 in the page).

This makes it difficult to link the result of the re-run to the original failure, which is leading to occasional false-positives from our CI system, which is a minor annoyance.

I am not sure that this is actually a bug, but I thought I woudl raise it.

I don't know much about Behat internals. Issue #154 might be related? If this is not considered to be a Behat probelm we can probably change our CI scripts which handle the test results to deal with our issue. Thank you for considering this.

**Comment by uuf6429** [comment]
@timhunt just to be clear..

>  Suppose random.html is a pointless page which displays 1 or 2 at random, and we test it like this.

The fact that it is random doesn't relate to the problem you are describing, right?

Also, you don't have to have a browser(or even the dummy html file) for this to happen, correct? (I'm just trying to exclude the unrelated points)

I tried what you described, and the output in the console looked fine, but I noticed the problem you mentioned in the junit output.
(side note: apparently we have some PHP 8.3 deprecation problems in the junit generator too)

**Comment by timhunt** [comment]
Right, random not essential. It was just the only think I could think of to get minimal steps to reproduce.

I think "An unambiguous solution would be to generate a scenario name based on the example values in all cases." is what I would expect. In my brain at least, "Unreliable test #2" is is "Unreliable test" run with the second set of Example data, whether or not the first set of data was executed as part of this run or not.

For me, it is a fairly fundamental princple of automated testing that each testcase should work the same, irrespective of any other tests that may or may not be being run. (And, certainly, each testcase works the same - naming is a relatively minor matter, but I still feel it should follow the same principle. But, that is just my opinion.)

Anyway, thank you very much for thinking about this.

**Comment by barryswaisland-eagleeye** [comment]
I think this is a great idea... I'd like to ask a couple of question on this... I understand this correctly given the scenario outline from above (with a slight change:
```
  Scenario Outline: Unreliable test
    When I load "random.html"
    Then I should see "<expected>"
    Examples:
      | expected |
      | a        |
      | b        |
```
Would this then report the scenario title appear in pretty/junit/html output as: `Unreliable test <expect:a>` or similar?

Could one use `bin/behat --name="Unreliable test <expect:a>"` to then run the scenario?

And lastly I tried to do what the referenced [issue 154](https://github.com/Behat/Behat/issues/154) indicates and doesn't seem to work for me, given `Scenario Outline: Unreliable test` is on line 100 of `Unreliable.feature` should one currently be able to use `bin/behat Unreliable.feature:100:1` to then run the first example of this scenario?

**Comment by acoulton** [comment]
@barryswaisland-eagleeye the work on this has been done in Behat/Gherkin but I don't think the new methods that added are actually used by Behat yet.

In future:

> Would this then report the scenario title appear in pretty/junit/html output as: Unreliable test <expect:a> or similar?
>
> Could one use bin/behat --name="Unreliable test <expect:a>" to then run the scenario? And if there were more columns could one still use a single column given that the values in that column are unique?

It would report the scenario title as "Unreliable test #1". If you wanted to include one or more of the column values you would explicitly include placeholders in the outline title e.g. `Scenario Outline: Unreliable test for <expected>` which would give you names of "Unreliable test for a #1" and "Unreliable test for b #2" in your example.

I'm fairly sure you will also be able to use the `--name` filter to run just that scenario.

> And lastly I tried to do what the referenced https://github.com/Behat/Behat/issues/154 indicates and doesn't seem to work for me, given Scenario Outline: Unreliable test is on line 100 of Unreliable.feature should one currently be able to use bin/behat Unreliable.feature:100:1 to then run the first example of this scenario?

The `Unreliable feature:100:1` syntax was suggested in that issue, but the actual syntax implemented is just to provide the line number of the example row.  So if I put line numbers on your example:

```
100 |  Scenario Outline: Unreliable test
102 |    When I load "random.html"
103 |    Then I should see "<expected>"
104 |    Examples:
105 |      | expected |
106 |      | a        |
107 |      | b        |
```

* `bin/behat unreliable.feature:100` will run all examples of that outline
* `bin/behat unreliable.feature:106` will run just the `a` example (the first one in the table)
* `bin/behat unreliable.feature:107` will run just the `b` example (the second one in the table)
